### ISSUE TITLE
~ServerApp never called

### ISSUE THREAD

**Issue by dbenage-cx** [freeorion/freeorion#1330]
### Environment
* **FreeOrion Version**: 0.4.6+
* **Operating System**: Fedora 25
* **Graphic card used**: N/A
* **Fetched as** 
  * Compiled from source b7ac413 

When resigning or exiting, `ServerApp::~ServerApp` never appears to execute.

Scenario 1 ("resign"):
*  `freeorion -q`
* Menu -> Resign

Scenario 2 ("wm_close"):
* `freeorion -q`
* Close via window (The "X" typically at the top right)

In neither scenario is there a log message [`Server::~Server`](https://github.com/freeorion/freeorion/blob/master/server/ServerApp.cpp#L161)

In scenario 2, `CleanupAIs()` is not called.  Last server log is from [`ServerFSM.cpp:241`](https://github.com/freeorion/freeorion/blob/master/server/ServerFSM.cpp#L241)
This sometimes results in ai clients and/or server remaining after exiting the game.

[logs_backtrace.zip](https://github.com/freeorion/freeorion/files/782706/logs_backtrace.zip)

Possibly related #1092
(optional for release?)

**Comment by geoffthemedio** [comment]
Try flushing the logger after the DebugLogger call? (I assume there's a way to do so though can't easily search code right now)

**Comment by geoffthemedio** [comment]
But there might need to be a delay added to the client after the close window button is clicked during which it waits for the server to respond to a shutdown message and end itself before forcibly shutting it down.

**Comment by dbenage-cx** [comment]
I had added a `std::cout`, without any output.
This first caught my attention during logger rework attempt.  I had [this commit](https://github.com/freeorion/freeorion/commit/a3a4d9907760960b65be73718dce3317ed324ebe), to (hopefully) overcome the windows crashes.

**Comment by geoffthemedio** [comment]
std::cout with a flush?

**Comment by dbenage-cx** [comment]
```
diff --git a/server/ServerApp.cpp b/server/ServerApp.cpp
index 4bf01db..c04f6e7 100644
--- a/server/ServerApp.cpp
+++ b/server/ServerApp.cpp
@@ -158,7 +158,9 @@ ServerApp::ServerApp() :
 }
 
 ServerApp::~ServerApp() {
+    std::cout << "ServerApp::~ServerApp" << std::endl;
     DebugLogger() << "ServerApp::~ServerApp";
+    BOOST_LOG_TRIVIAL(debug).flush();
     m_python_server.Finalize();
     CleanupAIs();
     delete m_fsm;
```

Neither message outputs in either scenario.
Edit: or `std::cout << "..." << std::flush;` (thought you meant logger)

**Comment by geoffthemedio** [comment]
I did mean logger, but you referred to std::cout in a previous comment.

Put a std::flush after the std::endl

**Comment by Vezzra** [comment]
> Possibly related #1092
(optional for release?)

If it's related to #1092 (which sounds plausible) I lean toward considering it something that should be addressed for 0.4.7. Tagging as "v0.4.7" for now.

**Comment by dbenage-cx** [comment]
Seems `std::endl` is basically `\n`+`std::flush`, I did try it anyway without success.
